# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E (от 0 до 30000)'
        elif 30001 <= income <= 50000:
            return 'D (от 30001 до 50000)'
        elif 50001 <= income <= 200000:
            return 'C (от 50001 до 200000)'
        elif 200001 <= income <= 1000000:
            return 'B (от 200001 до 1000000)'
        elif income >= 1000001:
            return 'A ( от 1000001 и выше)'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(data):
    try:
        if 'автом' in data:
            return 'операции с автомобилем'
        elif 'жил' in data or 'недвиж' in data:
            return 'операции с недвижимостью'
        elif 'свад' in data:
            return 'проведение свадьбы'
        elif 'образов' in data:
            return 'получение образования'
    except:
        return 'нет категории'

In [40]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)
data['purpose_category'].value_counts()

операции с недвижимостью    10751
операции с автомобилем       4279
получение образования        3988
проведение свадьбы           2313
Name: purpose_category, dtype: int64

### Шаг 3. Исследуйте данные и ответьте на вопросы

_

**Для начала посмотрим на все вещественные и целочисленные значения `data` и проведём валидацию соотносимых данных и столбцов.**

In [22]:
data.describe().round(2)

children  days_employed  dob_years  education_id  family_status_id  \
count  21331.00       21331.00   21331.00      21331.00          21331.00   
mean       0.47       67148.97      43.28          0.82              0.97   
std        0.75      139281.14      12.58          0.55              1.42   
min        0.00          24.14       0.00          0.00              0.00   
25%        0.00        1024.53      33.00          1.00              0.00   
50%        0.00        1996.73      42.00          1.00              0.00   
75%        1.00        5337.76      53.00          1.00              1.00   
max        5.00      401755.40      75.00          4.00              4.00   

           debt  total_income  
count  21331.00      21331.00  
mean       0.08     165342.75  
std        0.27      98313.99  
min        0.00      20667.00  
25%        0.00     107507.00  
50%        0.00     142594.00  
75%        0.00     195842.00  
max        1.00    2265604.00

**Выводы по столбцам:**

_

**children** (количество детей) - всё в порядке, несоответствий не обнаружено.

_

**days_employed** (общий трудовой стаж в днях) - **найдено несоответствие**. 

  Значение `max` слишком велико, не является возможным. Чтобы убедиться в этом поделим максимальное значение столбца `days_employed` на количество дней в году, 401755 / 365, это более **1100** лет трудового стажа. Значения `mean` и стандарное отклонение `std` так же выходят за рамки человеческой жизни. Если допустить, что некоторые аномальные значения указаны в часах, а не в днях, то поделив 401755 на 24, получается 16 739 дней, что является почти 45 годам трудового стажа, уже на правду. Так как не все значения в столбце указаны в часах, нам необходимо определить временные рамки, если значение выходит за них мы будем переводить часы в дни, остальное оставляем без изменений. Трудовой договор в России можно заключать с 16 лет, верхней границей будет максимальный возраст нашего датасета, это 75 лет. Получается 59 полных лет, с текущего дня прошло 21 548 дней, округлим до 21550.**

    *p.s. В данном случае мы не будем разделять верхнюю границу возраста для мужчин и женщин т.к. максимальное значение возвраста для женщин в нашем датасете = 74.*

_

**dob_years** - (возраст клиента в годах) **найдено несоответсвие.**

Минимальное значение не может быть равно нулю. Эти пропуски будем заполнять средним значением по `income_type`

_

**total_income** - (ежемесячный доход) **всё в порядке**, грубых несоответствий не обнаружено, **но** следует проверить эту строчку с максимальным значением на тип занятости.

###  Закончим предобработку данных и ответим на вопросы

**Исправим значения столбца `days_employed`**

In [23]:
#Отбираем некорректные через метод `loc` и сразу заменяем их разделив на 24

data.loc[data['days_employed'] > 21550, 'days_employed'] = data.loc[data['days_employed'] > 21550, 'days_employed'] / 24

#Проверяем результат
data.head(5)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0   14177.753002         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose     total_income_category  \
0               покупка жилья  B (от 200001 до 1000000)   
1     приобретение автомобиля    C (от 50001 до 200000)   
2               покупка жилья    C (от 50001 до 200000)   
3  дополнительное образование  B (от 200001 до 1000000)   
4             сыграть свадьбу    C (от 50001 до 200000)   

           purpose_category  
0  операции с недвижимостью  
1    операции с автомобилем  
2  операции с недвижимостью  
3     получение образования  
4        проведение свадьбы

In [24]:
data.describe()

children  days_employed     dob_years  education_id  \
count  21331.000000   21331.000000  21331.000000  21331.000000   
mean       0.474005    4597.015637     43.277952      0.816980   
std        0.752378    5350.743938     12.575300      0.549282   
min        0.000000      24.141633      0.000000      0.000000   
25%        0.000000    1024.525170     33.000000      1.000000   
50%        0.000000    1996.732855     42.000000      1.000000   
75%        1.000000    5337.759969     53.000000      1.000000   
max        5.000000   18388.949901     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21331.000000  21331.000000  2.133100e+04  
mean           0.974497      0.081196  1.653427e+05  
std            1.421787      0.273143  9.831399e+04  
min            0.000000      0.000000  2.066700e+04  
25%            0.000000      0.000000  1.075070e+05  
50%            0.000000      0.000000  1.425940e+05  
75%            1.000000      0.000000  1.958420e+05  
max            4.000000      1.000000  2.265604e+06

**Значения трудового стажа приведены к нормальным показателям**

_

**Добавим пропущенные значения в `dob_years`**

In [25]:
# С помощью цикла перебираем все уникальные значения в столбце income_type
for year in data['income_type'].unique():
    # Находим среднее значение dob_years для каждого типа занятости
    mean_value = data.loc[data['income_type'] == year, 'dob_years'].mean()
    # Находим dob_years равные нулю и типу занятости как у временной переменной year, 
    # меняем её значение на среднее по типу занятости
    data.loc[(data['dob_years'] == 0) & (data['income_type'] == year), 'dob_years'] = mean_value

In [26]:
#Проверяем результат
data.loc[data['dob_years'] == 0, 'dob_years'].count()

0

_

**Посмотрим на строку с максимальным значением из столбца `total_income`**

In [27]:
display(data[data['total_income'] == data['total_income'].max()])

children  days_employed  dob_years education  education_id  \
12412         0    1477.438114       44.0    высшее             0   

         family_status  family_status_id gender income_type  debt  \
12412  женат / замужем                 0      M   компаньон     0   

       total_income       purpose   total_income_category  \
12412       2265604  ремонт жилью  A ( от 1000001 и выше)   

               purpose_category  
12412  операции с недвижимостью

**Видим мужчину 44 лет с высшим образованием, не женат, тип занятости - компаньон. Высокий ежемесячный доход с такими показателями считаю допустимым.**

_

**После предобработки проверим тип данных в столбцах**

In [28]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21331 entries, 0 to 21524
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   children               21331 non-null  int64  
 1   days_employed          21331 non-null  float64
 2   dob_years              21331 non-null  float64
 3   education              21331 non-null  object 
 4   education_id           21331 non-null  int64  
 5   family_status          21331 non-null  object 
 6   family_status_id       21331 non-null  int64  
 7   gender                 21331 non-null  object 
 8   income_type            21331 non-null  object 
 9   debt                   21331 non-null  int64  
 10  total_income           21331 non-null  int64  
 11  purpose                21331 non-null  object 
 12  total_income_category  21331 non-null  object 
 13  purpose_category       21331 non-null  object 
dtypes: float64(2), int64(5), object(7)
memory usage: 2.4+ 

**Столбец  `days_employed` определяется как float64, а должен быть целочисленным (int), исправим это с помощью метода `astype`, в остальном всё в порядке, пропусков нет.**

In [29]:
data['days_employed'] = data['days_employed'].astype(int)

In [30]:
#Проверяем результат
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21331 entries, 0 to 21524
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   children               21331 non-null  int64  
 1   days_employed          21331 non-null  int64  
 2   dob_years              21331 non-null  float64
 3   education              21331 non-null  object 
 4   education_id           21331 non-null  int64  
 5   family_status          21331 non-null  object 
 6   family_status_id       21331 non-null  int64  
 7   gender                 21331 non-null  object 
 8   income_type            21331 non-null  object 
 9   debt                   21331 non-null  int64  
 10  total_income           21331 non-null  int64  
 11  purpose                21331 non-null  object 
 12  total_income_category  21331 non-null  object 
 13  purpose_category       21331 non-null  object 
dtypes: float64(1), int64(6), object(7)
memory usage: 2.4+ 

_

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Для ответа на этот вопрос нам необходимо создать сводную таблицу, в которой выделить три группы по количеству детей, и сравнить их показатели просроченной задолжности. Распределить группы следующим образом:
- 1) Нет детей
- 2) Есть дети
- 3) Многодетная (3 и более детей)

In [31]:
# Создадим функцию, которая будет возвращать соответвующуе название группы
def family_category(children):
    if 1 <= children <= 2:
        return 'Есть дети'
    if children >= 3:
        return 'Многодетная'
    return 'Нет детей'

In [32]:
# Проверим работу функции
print(family_category(3))

Многодетная


In [33]:
# Для создания сводной таблицы добавим результаты функции family_category
data['family_category'] = data['children'].apply(family_category)

In [59]:
# Создадим сводную таблицу методом pivot_table() ** Updated
debt_family = data.pivot_table(index = ['children'], values = 'debt', aggfunc = {'sum', 'count', 'mean'})
display(debt_family)

count      mean     sum
children                           
0         14091.0  0.075438  1063.0
1          4808.0  0.092346   444.0
2          2052.0  0.094542   194.0
3           330.0  0.081818    27.0
4            41.0  0.097561     4.0
5             9.0  0.000000     0.0

**Вывод:** Да, зависимость есть. Заёмщики не имеющие детей отдадут кредит с большей вероятностью, доля должников в этой категории составляет 7,5%. Более высокий риск просрочить платеж имеют те заёмщики, у кого 1 или 2 ребёнка, здесь доля должников составляет 9,2 % и 9,4%. Многодетных заёмщиков в нашем случае не стоит брать в рассчёт, так как выборка очень маленькая.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

**Для ответа на этот вопрос будем действовать аналогичным образом. Создадим сводную таблицу, где выделим все пять групп и сравним долю должников в каждой.**

In [35]:
# Создадим сводную таблицу методом pivot_table()
debt_family_status = data.pivot_table(index='family_status', columns='debt', values='gender', aggfunc='count')

# Добавим столбцы для определения наличия просроченной задолженности
debt_family_status.columns = ['no_debt', 'have_debt']

# Считаем долю должников, разделим всех должников на общее количество каждой группы
debt_family_status['debtor_share'] = debt_family_status['have_debt'] / (debt_family_status['have_debt'] + debt_family_status['no_debt'])

# Для наглядности переведём значение в проценты
debt_family_status['debtor_share'] = debt_family_status['debtor_share'].map('{:.2%}'.format)

# Сортируем по столбцу доля лолжников по убыванию
debt_family_status.sort_values(by='debtor_share', ascending=False)

no_debt  have_debt debtor_share
family_status                                         
Не женат / не замужем     2523        273        9.76%
гражданский брак          3749        385        9.31%
женат / замужем          11334        927        7.56%
в разводе                 1105         84        7.06%
вдовец / вдова             888         63        6.62%

**Вывод:** Да, зависимость есть. На выборке по семейному положению нам удалось выделить более и менее рисковые группы. К более рисковым заёмщикам относятся не замужние и не женатые заёмщики, а так же те, кто не зарегистрировал свой бранчный союз официально, доля должников в двух этих группах самая высокая, 9,76% и 9,31%. Заёмщики состоящие в официальном браке имеют меньшую склонность к просрочке платежей, доля должников здесь 7,56%. Самыми ответсвенными оказались одинокие люди, вдовцы и вдовы всего 6,62%, находящиеся в разводе - 7,06%. 

_

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

**Создадим сводную таблицу по созданным выше категориям дохода и сравним долю должников.**

In [36]:
# Создадим сводную таблицу методом pivot_table()
income_status = data.pivot_table(index='total_income_category', columns='debt', values='gender', aggfunc='count')

# Добавим столбцы для определения наличия просроченной задолженности
income_status.columns = ['no_debt', 'have_debt']

# Считаем долю должников, разделим всех должников на общее количество каждой группы
income_status['debtor_share'] = income_status['have_debt'] / (income_status['have_debt'] + income_status['no_debt'])

# Для наглядности переведём значение в проценты
income_status['debtor_share'] = income_status['debtor_share'].map('{:.2%}'.format)

# Сортируем по столбцу доля лолжников по убыванию
income_status.sort_values(by='debtor_share', ascending=False)

no_debt  have_debt debtor_share
total_income_category                                    
E (от 0 до 30000)              20          2        9.09%
C (от 50001 до 200000)      14568       1353        8.50%
A ( от 1000001 и выше)         23          2        8.00%
B (от 200001 до 1000000)     4660        354        7.06%
D (от 30001 до 50000)         328         21        6.02%

**Вывод:** Да, зависимость есть. В этой выборке стоит сравнивать только категории B и C, так как по остальным группам результаты не являются репрезентативными. Люди с доходом от 50 000 до 200 000 т.р. более склонны пропускать платежи, чем те, чей ежемесячный доход составляет от 200 000 т.р. до 1 млн. руб. Доля должников в группе С состовляет 8,5%, группа B - 7,06%. Рекомендация: Сделать отдельный анализ по этой выборке, разбить на более мелкие группы, уменьшить шаг до  20-30 т.р. для дохода до 200 т.р. Но этого понадобится в несколько раз больше данных, чем было представлено. 

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [37]:
# Создадим сводную таблицу методом pivot_table()
purpose_status = data.pivot_table(index='purpose_category', columns='debt', values='gender', aggfunc='count')

# Добавим столбцы для определения наличия просроченной задолженности
purpose_status.columns = ['no_debt', 'have_debt']

# Считаем долю должников, разделим всех должников на общее количество каждой группы
purpose_status['debtor_share'] = purpose_status['have_debt'] / (purpose_status['have_debt'] + purpose_status['no_debt'])

# Для наглядности переведём значение в проценты
purpose_status['debtor_share'] = purpose_status['debtor_share'].map('{:.2%}'.format)

# Сортируем по столбцу доля лолжников по убыванию
purpose_status.sort_values(by='debtor_share', ascending=False)


no_debt  have_debt debtor_share
purpose_category                                         
операции с автомобилем       3879        400        9.35%
получение образования        3619        369        9.25%
проведение свадьбы           2130        183        7.91%
операции с недвижимостью     9971        780        7.26%

**Вывод:** Взглянув на выборку по целям кредита тоже можно тоже выделить разную степень ответсветвенности заёмщика. В группу риска попадают те заёмщики, кто собирается купить автомобиль, у них самая высокая доля просрочки платежа, 9,35%. Так же очень близко к ним расположились те, кто хочет получить образование на заёмные средства, доля просрочки 9,25%. Чуть более ответсвенны те, кто хочет сыграть свадьбу на кредитные деньги, доля просроченных платежей составляет 7,91%, похожие значения мы уведили выше, у заёмщиков состоящих в браке. Самыми ответсвенными заёмщиками в этой выборке оказались люди которые хотят приобрести жильё, тут доля просроченных платежей состовляет 7,26%. Такие показатели могут быть связаны с развитием в последние годы различных государсвенных программ, например "Льготная ипотека", "Дальневосточная ипотека", "Семейная ипотека". 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
- Сокрытие информации
- Технические проблемы
- Неверный ввод данных

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Для количественных значений медиана лучше, т.к. в нашем случае был большой разброс значений, а медиана более устойчива к этому фактору.

### Шаг 4: общий вывод.

Напишите ваш общий вывод.

**Резюмируем: Полученные данные позволяют выделить несколько групп более благонадежных заёмщиков.**

- Желающие взять ипотечный кредит, доля просрочки 7,26%
- Женатые без детей, доля просрочки 7,56%
- Разведенные, вдовы и вдовцы, доля просрочки 6,02%
- с высоким уровнем дохода (от 200001 до 1 000 000), доля просрочки 7,06%

**С другой стороны можем выделить несколько групп с повышенным риском к образованию просроченной задолженности.**

- Холостые или состоящие в гражданском браке, их доля 9,76% и 9,31%
- Заёмщики с имеющие одного или двух детей, здесь доля должников составляет 9,2 % и 9,4% 
- Желающие взять автокредит или оплатить образование, доля должников 9,35% и 9,25%